src/load.rs

```diff
@@ -121,65 +140,69 @@ impl Loader {
-            None => false,
-            Some("gcc") => false,
-            Some("msvc") => true,
-            Some(other) => bail!("invalid deps attribute {:?}", other),
-        };
-        let pool = lookup("pool");
-
-        let rspfile_path = lookup("rspfile");
-        let rspfile_content = lookup("rspfile_content");
-        let rspfile = match (rspfile_path, rspfile_content) {
-            (None, None) => None,
-            (Some(path), Some(content)) => Some(RspFile {
-                path: std::path::PathBuf::from(path),
-                content,
-            }),
-            _ => bail!("rspfile and rspfile_content need to be both specified"),
+            let cmdline = lookup("command");
+            let desc = lookup("description");
+            let depfile = lookup("depfile");
+            let deps = match lookup("deps").as_deref() {
+                None => None,
+                Some("gcc") => Some("gcc".to_string()),
+                Some("msvc") => Some("msvc".to_string()),
+                Some(other) => bail!("invalid deps attribute {:?}", other),
+            };
+            let pool = lookup("pool");
+
+            let rspfile_path = lookup("rspfile");
+            let rspfile_content = lookup("rspfile_content");
+            let rspfile = match (rspfile_path, rspfile_content) {
+                (None, None) => None,
+                (Some(path), Some(content)) => Some(RspFile {
+                    path: std::path::PathBuf::from(path),
+                    content,
+                }),
+                _ => bail!("rspfile and rspfile_content need to be both specified"),
+            };
+            let hide_success = lookup("hide_success").is_some();
+            let hide_progress = lookup("hide_progress").is_some();
+
+            build.cmdline = cmdline;
+            build.desc = desc;
+            build.depfile = depfile;
+            build.deps = deps;
+            build.rspfile = rspfile;
+            build.pool = pool;
+            build.hide_success = hide_success;
+            build.hide_progress = hide_progress;
         };
-        let hide_success = lookup("hide_success").is_some();
-        let hide_progress = lookup("hide_progress").is_some();
-
-        build.cmdline = cmdline;
-        build.desc = desc;
-        build.depfile = depfile;
-        build.parse_showincludes = parse_showincludes;
-        build.rspfile = rspfile;
-        build.pool = pool;
-        build.hide_success = hide_success;
-        build.hide_progress = hide_progress;
-
-        self.graph.add_build(build)
+
+        self._graph.borrow_mut().add_build(build)
     }
 
     fn read_file(&mut self, id: FileId) -> anyhow::Result<()> {
-        let path = self.graph.file(id).path().to_path_buf();
-        let bytes = match trace::scope("read file", || scanner::read_file_with_nul(&path)) {
-            Ok(b) => b,
-            Err(e) => bail!("read {}: {}", path.display(), e),
+        let path = self.graph().file(id).path().to_path_buf();
```

Comment:
Note to self: I guess this could've been avoided if `bail!` had been done inside `FileLoader`'s `read_file` instead?

Reply:
Moved it inside `read_file`.